eth-contracts/contracts/Governance.sol

```diff
@@ -0,0 +1,294 @@
+
+    // Require proposal votingPeriod is still active.
+    uint256 startBlockNumber = slashProposals[_proposalId].startBlockNumber;
+    uint256 endBlockNumber = startBlockNumber + votingPeriod;
+    require(
+      block.number > startBlockNumber && block.number <= endBlockNumber,
+      "Proposal votingPeriod has ended"
+    );
+
+    // Require vote is not None.
+    require(_vote != Vote.None, "Cannot submit None vote");
+
+    // Record previous vote.
+    Vote previousVote = slashProposals[_proposalId].votes[voter];
+
+    // Will override staker's previous vote if present.
+    slashProposals[_proposalId].votes[voter] = _vote;
+
+    /** Update voteMagnitudes accordingly */
+
+    // New voter (Vote enum defaults to 0)
+    if (previousVote == Vote.None) {
+      if (_vote == Vote.Yes) {
+        slashProposals[_proposalId].voteMagnitudeYes += voterStake;
+      } else {
+        slashProposals[_proposalId].voteMagnitudeNo += voterStake;
+      }
+      slashProposals[_proposalId].numVotes += 1;
+    }
+    // Repeat voter
+    else {
+      if (previousVote == Vote.Yes && _vote == Vote.No) {
+        slashProposals[_proposalId].voteMagnitudeYes -= voterStake;
+        slashProposals[_proposalId].voteMagnitudeNo += voterStake;
+      }
+      else if (previousVote == Vote.No && _vote == Vote.Yes) {
+        slashProposals[_proposalId].voteMagnitudeYes += voterStake;
+        slashProposals[_proposalId].voteMagnitudeNo -= voterStake;
+      }
+      // If _vote == previousVote, no changes needed to vote magnitudes.
+    }
+
+    emit SlashProposalVoteSubmitted(
+      _proposalId,
+      voter,
+      _vote,
+      voterStake,
+      previousVote
+    );
+  }
+
+  function evaluateSlashProposalOutcome(uint256 _proposalId)
+  external returns (Outcome proposalOutcome)
+  {
+    require(_proposalId <= lastProposalId && _proposalId > 0, "Must provide valid non-zero _proposalId");
+
+    // Require msg.sender is active Staker.
+    Staking stakingContract = Staking(registry.getContract(stakingProxyOwnerKey));
+    require(
+      stakingContract.totalStakedForAt(
+        msg.sender, slashProposals[_proposalId].startBlockNumber
+      ) > 0,
+      "Caller must be active staker with non-zero stake."
+    );
+
+    // Require proposal votingPeriod has ended.
+    uint256 startBlockNumber = slashProposals[_proposalId].startBlockNumber;
+    uint256 endBlockNumber = startBlockNumber + votingPeriod;
+    require(
+      block.number > endBlockNumber,
+      "Proposal votingPeriod must end before evaluation."
+    );
+
+    // Calculate outcome
+    Outcome outcome;
+    if (slashProposals[_proposalId].numVotes < votingQuorum) {
+      outcome = Outcome.Invalid;
+    }
+    else if (
+      slashProposals[_proposalId].voteMagnitudeYes >= slashProposals[_proposalId].voteMagnitudeNo
+    ) {
+      outcome = Outcome.Yes;
+
+      // Slash target's stake.
+      stakingContract.slash(
```

Comment:
(non-blocking for v0). This has to become DelegateManager OR the total slash amount has to be restricted by SP deployer stake - we can chat about that offline

Reply:
no, a slash action will NOT change the staking multiplier - that would reduce everyone's balance, not just the target of slashing. it will change the staked amount in Staking.sol / SpFactory/DelegateManager - so this is quite a complex operation

Reply:
@hareeshnagaraj yes sorry that's what i meant. thanks for clarifying